R/LSgameofchance.R

```diff
@@ -15,272 +15,146 @@
-  for (i in 1:simulation){
-    # copy the trials players need to succeed and update after each trial
-    record_trial <- k # each row stands for simulations
-    # keep running the game until one record of the trial reaches 0
-    while (prod(record_trial) != 0){
-      record_trial <- record_trial - t(rmultinom(1,1,p)) #updating the recorded result
-    }
-    # transfer the record of trial to the record of game
-    record_game[i] <- which(record_trial == 0, arr.ind = TRUE)[1,"col"]
-  }
-  
-  # estimated probability that each player wins
-  p_simulated <- vector()
-  for (i in 1:length(p)){
-    p_simulated[i] <- length(which(record_game == i))/simulation 
-  }
-  
-  p_cumulative <- matrix(0, nrow = length(p), ncol = simulation)
-  p_cumulative[record_game[1],1] <- 1
-  for (i in 1:length(p)){    
-    for (j in 2:simulation){
-      
-      if (record_game[j] == i){
-        p_cumulative[i, j] <- (p_cumulative[i, j-1]*(j-1)+1)/j
-      }else{
-        p_cumulative[i, j] <- p_cumulative[i, j-1]*(j-1)/j
-      }
-    }
-  }
-  
-  ## Calculating the probability using negative multinomial distribution and resursive function
-  p_calculated <- 0
-  for (l in 1:(prod(k)/k[1])){
-    p_calculated <- p_calculated + exp(MGLM::dnegmn(Y = combinations(k)[l,2:length(k)],
-                                              beta = combinations(k)[1], prob = p[2:length(p)]))
-  }
-  
-  # calculating difference between the simulated and the calculated probability of player 1 winning
-  dif <- abs(p_simulated - p_calculated)
-  return (list(p_simulated, p_calculated, dif, p_cumulative))
-}
-
-
-#l <- compare_function2(c(1,1,1), 2, c(0.2,0.4,0.4), 10)
-#l
-
-combinations <- function(k){
-  if (length(k) == 3){ # the number of players
-    #k <- c(1,2,3)
-    num <- prod(k)/k[1] #total number of sets
-    #sets <- vector(mode = "list", length = num)
-    sets <- matrix(nrow = num, ncol = length(k))
-    for (i in 0:(k[2]-1)){
-      for (j in 0:(k[3]-1)){
-        sets[i*k[3]+j+1,] <- c(k[1],i,j)
-      }
-    }
-    return(sets)
-  }else{
-    #k <- c(1,2,3,4)
-    #x1 <- sets
-    x1 <- combinations(k[1:(length(k) - 1)]) # k without k[length(k)]
-    x2 <- k[length(k)]  
-    y <- matrix(ncol = length(k), nrow = prod(k)/k[1])
-    for (i in 1:length(x1[,1])){
-      for (j in 1:k[length(k)]){
-        y[(i-1)*(k[length(k)])+j,] <- append(x1[i,],j-1)
-      }
-    }
-    return(y)
-  }
-} 
-
-##
 LSgameofchance   <- function(jaspResults, dataset, options, state = NULL){
 
-  #ready <- (length(options$variables) > 0)
+  # input values
+  nPlayers  <- length(options[["players"]])
+  probWin   <- sapply(options[["players"]], function(p)p$values[[1]])
+  xPoints   <- sapply(options[["players"]], function(p)p$values[[2]])
+  winPoints <- options[["winPoints"]]
+  nSims     <- options[["nSims"]]
   
+  # normalizing the probability of wining:
+  probWin   <- probWin / sum(probWin) # normalit
```

Comment:
renaming the arguments for easier orientation in the code

Reply:
and I should've used `p[["values"]]` here as well

Reply:
if you want to go all the way with this you could also write a custom class :)

Reply:
R6 JASP classes ftw :D